### DISCUSSION TITLE
StateNotifier listening to another provider

### DISCUSSION THREAD

**Discussion by SamiAlsubhi** [rrousselGit/riverpod#1538] [Q&A]
I have StateNotifier Providers that I need them to listen to other providers but without rebuilding/recreating the StateNotifier Provider.

The examples below are functionally pointless and they are just for demoing.

So I came up with some ways.
First:

```dart
final provider = StateProvider<String>((ref) {
  ref.listen<String>(fromProvider, (_, val) {
    ref.read(toProvider.notifier).setVal(val);// sometimes error here:
   //Err: Cannot use ref functions after the dependency of a provider changed but before the provider rebuilt
  });
  return ref.watch(toProvider);
});

final toProvider = StateProvider<ToNotifier,String>((ref)=>ToNotifier(ref))

class ToNotifier extends StateNotifier<String> {
   Ref _ref;
   ToNotifier(this._ref):super("");

   void setVal(String val){
       state = val
   }
}
```
Second:
```dart

final provider = StateProvider<ToNotifier,String>((ref)=>ToNotifier(ref))

class ToNotifier extends StateNotifier<String> {
   Ref _ref;
   ToNotifier(this._ref):super(""){
       ref.listen<String>(fromProvider, (_, val) {
           if(!mounted) return;
           setVal(val); //no above error but I am calling ref.listen asynchronously
       });
   }

   void setVal(String val){
       state = val
   }
}
```

Third:
```dart

final provider = StateProvider<ToNotifier,String>((ref)=>ToNotifier(ref))

class ToNotifier extends StateNotifier<String> {
   Ref _ref;
   late void Function() removeListner;
   ToNotifier(this._ref):super(""){
       removeListner = ref.read(fromProvider).addListner((val) {
           if(!mounted) return;
           setVal(val); 
       }, fireImmediately: false);
   }

   void setVal(String val){
       state = val
   }

@override
  void dispose() {
    removeListner();
    super.dispose();
  }
}

```
Which one is safe to be used (apparently not the first one)?
Maybe if this is legal (which I hope it is) can be added to docs.
Or if this is totally not supported then I hope it gets implemented in the next version.

**Comment by rrousselGit** [comment]
Your first example seems odd. I don't see the purpose of the ref.listen, considering the ref.watch already takes care of that.

Out of 2 & 3, 2 is preferable. As it's possible for the StateNotifier instance to change, which 3 does not handle.

**Reply by SamiAlsubhi** [reply]
The idea here is that I do not need the `fromProvider` to rebuild StateNotifier.

I know The example I provided does not demonstrate a useful case.

To better explain; one use case I have is that The StateNotifier has an ongoing WebSocket connection. This connection requires to read device location from another provider. The StateNotifier's state holds the the WS incoming messages.

Therefore, the location provider doesn't interfere with the StateNotifier's state. It just provides data to the WS sink. That's why I need this.

Based on this information: is example 2 still the preferred way to do this?
